### ISSUE TITLE
Okta OpenID Connect Configuration causing target registration timeouts in AWS ECS

### ISSUE THREAD

**Issue by elliott-pope-smarsh** [micronaut-projects/micronaut-security#604]
### Steps to Reproduce

1. Create an HTTP API protected with Okta OpenID Connect Configuration and the `/health` endpoint enabled
`build.gradle`
```
implementation("io.micronaut.security:micronaut-security-oauth2")
implementation("io.micronaut.security:micronaut-security-jwt")
```
`application.yml`
```
micronaut:
  security:
    authentication: idtoken
    oauth2:
      clients:
        okta:
          client-id: <client id>
          client-secret: <client secret>
          openid:
            issuer: <issuer URL>
endpoints:
  health:
    enabled: true
    sensitive: false
    details-visible: AUTHENTICATED
```
2. Deploy that on a `t2.micro` instance on AWS ECS
3. Configure an Application Load Balancer to call the `/health` endpoint to register the target

### Expected Behaviour

Container should become healthy and target should be registered

### Actual Behaviour

Call times out and the target fails to register. The logs show 
```
[default-nioEventLoopGroup-1-2] ERROR io.micronaut.http.server.netty.RoutingInBoundHandler - Unexpected error occurred: Error instantiating bean of type [io.micronaut.security.oauth2.client.DefaultOpenIdProviderMetadata]: Read Timeout
```

After some discussion on the Gitter, I think that this call to Okta is blocking the `default` event loop which prevents the server from responding to requests. I was able to resolve the issue by configuring separate client and server event loops of size 10
```
micronaut:
  server:
    netty:
      worker:
        event-loop-group: server-event-loop
  netty:
    event-loops:
      server-event-loop:
        num-threads: 10
      client-event-loop:
        num-threads: 10
  http:
    client:
      event-loop-group: client-event-loop
```
but this shouldn't be necessary

### Environment Information

- **Operating System**: AWS ECS Optimized Linux
- **Micronaut Version:** 2.4.1
- **JDK Version:** 15

### Example Application

I'll work on a sample application so that this can be more easily tested and verified. The application that this issue is coming from is proprietary

**Comment by sdelamo** [comment]
Please, update a sample app if you can. 

Without doing changes to the event loop could you try also to increase Micronaut HTTP Client read timeout. 

`http.client.read-timeout`

**Comment by graemerocher** [comment]
increasing the read timeout is not the solution to this. Need to review why the okta config is blocking the event loop

**Comment by graemerocher** [comment]
Likely culprit https://github.com/micronaut-projects/micronaut-security/blob/ee9e964a70c4bf263a9cdc9bb5dc190777627fea/security-oauth2/src/main/java/io/micronaut/security/oauth2/client/OpenIdClientFactory.java#L93

**Comment by donarus** [comment]
Any progress on this ? Seems unresolved so far . I'm struggling with the identical issue.

Micronaut 3.3.0
ECS
Oauth over Keycloak
Health endpoint enabled

Running locally just fine .. running in ECS failing on this error

Is the cause known or would it be helpful to provide more derails?

**Comment by graemerocher** [comment]
@donarus does the workaround described in the issue report resolve the issue?

**Comment by donarus** [comment]
@graemerocher yes, it does, but as Elliot mentioned, it shouldn't be.

**Comment by ttimt96** [comment]
Encountered same issue on 3.4.2 and had to apply the workaround

**Comment by adamkobor** [comment]
While the proposed workaround probably solves the server related timeout issue, I think it introduces a new one for client calls. With Keycloak set up as our OIDC provider we are facing a weird issue that is super hard to reproduce: when a client call happens very close to the mentioned (blocking) one in `OpenIdClientFactory` after the application starts up, sometimes it will fail with the following error:

```
14:24:16.063 [DefaultDispatcher-worker-1] ERROR i.m.h.c.i.HttpClientIntroductionAdvice - Client [op.kaltura.SomeDeclarativeClient] received HTTP error response: Error instantiating bean of type  [io.micronaut.security.token.jwt.validator.JwtTokenValidator]
```

My idea for the fix was that the blocking call should happen before the server starts, so the only thing that seems to be able to fix it is this workaround (but I'm still not sure, have to keep an eye on our logs):

```kotlin
@Context
@Requires(notEnv = [Environment.TEST], bean = OpenIdClientConfiguration::class)
@Suppress("UNUSED")
class EagerOpenIdMetadataInitializer(private val metadata: OpenIdProviderMetadata) {

    private val logger = LoggerFactory.getLogger(EagerOpenIdMetadataInitializer::class.java)

    init {
        logger.info { "OpenIdProviderMetadata was eagerly initialized for the following issuer: ${metadata.issuer}" }
    }
}
```

Would it be possible to do something like this in `OpenIdClientFactory` by default?

**Comment by graemerocher** [comment]
eager initialization seems wise. PRs welcome

**Comment by sdelamo** [comment]
I think this should fix the issue: https://github.com/micronaut-projects/micronaut-security/commit/e142b12f19edbbbb5dde0bcdb5cfe9f70bb598b6

**Comment by sdelamo** [comment]
It will be released in Micronaut security 3.9.0
